### PR TITLE
Re-introduce "Leave out optional keys from /sync" change

### PR DESCRIPTION
**Pull request by anoadragon453**
Closes: https://github.com/matrix-org/synapse/issues/9941

Required some fixes due to merge conflicts with #6739, but nothing too hairy. The first commit is the same as the original (after merge conflict resolution) then two more for compatibility with the latest sync code.

cc @callahad

### CODE REVIEW

File: changelog.d/9919.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Omit empty fields from the `/sync` response.  Contributed by @deepbluev7.
```
Comment (richvdh) [inline changelog.d/9919.feature:1]: I think we should omit this.

File: tests/server_notices/test_resource_limits_server_notices.py [hunk 2]
```diff
@@ -321,7 +327,9 @@ def test_invite_with_notice(self):
 
         # Scan the events in the room to search for a message from the server notices
         # user.
-        events = channel.json_body["rooms"]["join"][room_id]["timeline"]["events"]
+        events = channel.json_body["rooms"][Membership.JOIN][room_id]["timeline"][
```
Comment (richvdh) [inline tests/server_notices/test_resource_limits_server_notices.py:330]: as a matter of general preference, I would avoid using symbolic constants like this in tests where possible. It's a chance to spot typos in the definition of the constant!

**Review by richvdh — APPROVED** [review]
lgtm modulo the duplicate changelog

File: tests/server_notices/test_resource_limits_server_notices.py (hunk 2 above)
Comment (anoadragon453) [reply tests/server_notices/test_resource_limits_server_notices.py:330]: That's a good point! I also did so when adding the tests for knocking, though that made a bit more sense when we needed to switch the prefix from unstable to stable.

But a good tactic to carry forward.

**Comment by erdnaxeli** [comment]
I just want you to notice that this can broke lib :grimacing: 

Actually the spec is not helpfull to know which fields will always be there and which are optional. Sometimes the field is required and this is specified, like in the response of `/login`:
>  m.homeserver | Homeserver Information | Required. Used by clients to discover homeserver information.

Sometimes the field is optional and this is specified, like in the description on `UnsignedData`:
> redacted_because | Event | Optional. The event that redacted this event, if any.

But often there is just nothing, and we don't know. The specs should specify at the start of the document a default behavior, and then remove the useless precision ("required" or "optional", depending on the default).

**Comment by ShadowJonathan** [comment]
All fields in `/sync` are optional, except next_batch, it's just that synapse has always returned them by default.

**Comment by erdnaxeli** [comment]
Well, as I was saying, the spec does not make it clear.

**Pull request merged**